src/Manipulators/Encode.php

```diff
@@ -15,30 +12,39 @@ class Encode extends BaseManipulator
      *
      * @param ImageInterface $image The source image.
      *
-     * @return ImageInterface The manipulated image.
+     * @return EncodedImageInterface The encoded image.
      */
-    public function run(ImageInterface $image): ImageInterface
+    public function run(ImageInterface $image): EncodedImageInterface
     {
         $format = $this->getFormat($image);
         $quality = $this->getQuality();
-        $driver = $image->driver();
-        $interlace = false;
+        $shouldInterlace = filter_var($this->getParam('interlace'), FILTER_VALIDATE_BOOLEAN);
 
         if ('pjpg' === $format) {
-            $interlace = true;
-
+            $shouldInterlace = true;
             $format = 'jpg';
         }
 
-        $image = (new ImageManager($driver))->read(
-            $image->encodeByExtension($format, $quality)->toString()
-        );
-
-        if ($interlace) {
-            $image = $this->interlace($image, $driver);
+        $encoderOptions = ['extension' => $format];
+        switch ($format) {
+            case 'avif':
+            case 'heic':
+            case 'tiff':
+            case 'webp':
+                $encoderOptions['quality'] = $quality;
+                break;
+            case 'jpg':
+                $encoderOptions['quality'] = $quality;
+                $encoderOptions['progressive'] = $shouldInterlace;
+                break;
+            case 'gif':
+            case 'png':
+                $encoderOptions['interlaced'] = $shouldInterlace;
+                break;
+            default:
         }
 
-        return $image;
+        return $image->encodeByExtension(...$encoderOptions);
```

Comment:
Due the latest changes on Intervention package. It was necessary to change the return type of this class, to return the encoded image instance instead just the image instance.